org.lflang.tests/src/org/lflang/tests/TestBase.java

```diff
@@ -426,48 +456,163 @@ private void generateCode(LFTest test) {
-                if (p.exitValue() == 0) {
-                    test.result = Result.TEST_PASS;
+            for (ProcessBuilder pb : pbList) {
+                var p = pb.start();
+                var stdout = test.execLog.recordStdOut(p);
+                var stderr = test.execLog.recordStdErr(p);
+                if (!p.waitFor(MAX_EXECUTION_TIME_SECONDS, TimeUnit.SECONDS)) {
+                    stdout.interrupt();
+                    stderr.interrupt();
+                    p.destroyForcibly();
+                    test.result = Result.TEST_TIMEOUT;
+                    return;
                 } else {
-                    test.result = Result.TEST_FAIL;
+                    if (p.exitValue() != 0) {
+                        test.result = Result.TEST_FAIL;
+                        return;
+                    }
                 }
             }
-
         } catch (Exception e) {
             test.result = Result.TEST_FAIL;
+            return;
+        }
+        test.result = Result.TEST_PASS;
+    }
+
+    /**
+     * Return a Mapping of federateName -> absolutePathToDockerFile.
+     * Expects the docker file to be two levels below where the source files are generated (ex. srcGenPath/nameOfFederate/*Dockerfile)
+     * @param test The test to get the execution command for.
+     */
+    private Map<String, Path> getFederatedDockerFiles(LFTest test) {
+        Map<String, Path> fedNameToDockerFile = new HashMap<>();
+        File[] srcGenFiles = test.fileConfig.getSrcGenPath().toFile().listFiles();
+        for (File srcGenFile : srcGenFiles) {
+            if (srcGenFile.isDirectory()) {
+                File[] dockerFile = srcGenFile.listFiles(new FileFilter() {
+                    @Override
+                    public boolean accept(File pathName) {
+                        return pathName.getName().endsWith("Dockerfile");
+                    }
+                });
+                assert dockerFile.length == 1;
+                fedNameToDockerFile.put(srcGenFile.getName(), dockerFile[0].getAbsoluteFile().toPath());
+            }
+        }
+        return fedNameToDockerFile;
+    }
+
+    /**
+     * Return the content of the bash script used for testing docker option in federated execution. 
+     * @param fedNameToDockerFile A mapping of federateName -> absolutePathToDockerFile
+     * @param testNetworkName The name of the network used for testing the docker option. 
+     *                        See https://github.com/lf-lang/lingua-franca/wiki/Containerized-Execution#federated-execution for more details.
+     */
+    private String getDockerRunScript(Map<String, Path> fedNameToDockerFile, String testNetworkName) {
+        StringBuilder shCode = new StringBuilder();
+        shCode.append("#!/bin/bash\n");
+        int n = fedNameToDockerFile.size();
+        shCode.append("pids=\"\"\n");
+        shCode.append(String.format("docker run --rm --network=%s --name=rti rti:test -i 1 -n %d &\n", testNetworkName, n));
```

Comment:
Would `ProcessBuilder` not be able to execute these commands and read their return value?

Reply:
To test federated execution, several processes need to run concurrently. The current implementation of the test framework does not allow that.

Reply:
I see. I guess the problem is that some commands do need to run sequentially (e.g., `docker build` and `docker run` for each federate) and some can (and should) run concurrently.

I think if we design this carefully enough, we could use something like Java thread pools to achieve concurrency. Not using bash means better compatibility with Windows, and Docker-enabled federated programs can really be a shortcut to getting federated execution to work on Windows. It's your choice though of course.